unity-renderer/Assets/DCLPlugins/ECS7/ECSComponents/Transform/Handler/ECSTransformHandler.cs

```diff
@@ -22,12 +22,43 @@ public ECSTransformHandler(IWorldState worldState, BaseVariable<Vector3> playerT
 
         public void OnComponentRemoved(IParcelScene scene, IDCLEntity entity)
         {
+            ECSTransformUtils.orphanEntities.Remove(entity);
+
+            // reset transform and re-parent to the scene
+            entity.gameObject.transform.ResetLocalTRS();
+            ECSTransformUtils.TrySetParent(scene, entity, SpecialEntityId.SCENE_ROOT_ENTITY);
+
+            // if entity has any parent
             if (entity.parentId != SpecialEntityId.SCENE_ROOT_ENTITY)
             {
+                // remove as child
+                if (scene.entities.TryGetValue(entity.parentId, out IDCLEntity parent))
+                {
+                    parent.childrenId.Remove(entity.entityId);
+                }
+
                 entity.parentId = SpecialEntityId.SCENE_ROOT_ENTITY;
-                ECSTransformUtils.SetParent(scene, entity, SpecialEntityId.SCENE_ROOT_ENTITY);
             }
-            ECSTransformUtils.orphanEntities.Remove(entity);
+
+            // if entity has any children
+            int childrenCount = entity.childrenId.Count;
+            if (childrenCount > 0)
+            {
+                for (int i = childrenCount - 1; i >= 0; i--)
+                {
+                    long childId = entity.childrenId[i];
+                    entity.childrenId.RemoveAt(i);
```

Comment:
Here we are affecting the same collection that is being traversed, doesn't this bring any problem?

Maybe it's better to avoid doing the `RemoveAt()` here and just do `entity.childrenId.Clean()` after the `for`?

Reply:
affecting the collection shouldn't matter since the list is being iterated backwards.
but you are totally right regarding performance, since it will be better just to clear the list after the for.
changed here :https://github.com/decentraland/unity-renderer/pull/2804/commits/6631a1cdfb212e14d4c5c688493442f8f3a3f715